File: 47.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
    vector<vector<int>> permuteUnique(vector<int>& nums) {
        vector<vector<int>> res;
        //sort(nums.begin(), nums.end());
        search(res, nums, 0);
        return res;
    }

    void search(vector<vector<int>> &res, vector<int> &nums, int pos) {
        if (pos == nums.size()) {
            // for(vector<int> t : res) {
            //     if(t == nums) return;
            // }
            res.push_back(nums);
            cout << "pos:" << pos << endl;
        }else {
            for(int i = pos; i < nums.size(); ++i) {
                int ifContinue = 0;
                for(int j = i + 1; j < nums.size(); ++j) {
                    if(nums[j] == nums[i]) ifContinue = 1;
                }
                if(ifContinue) continue;
                swap(nums, i, pos);
                //cout << "i:" << i << " pos:" << pos << endl;
                search(res, nums, pos + 1);
                swap(nums, i, pos);
                //while(nums[i+1] == nums[i]) ++i;
            }
        }
    }

    void swap(vector<int> &nums, int i, int j) {
        int t = nums[i];
        nums[i] = nums[j];
        nums[j] = t;
    }
};

File: 47.py
from typing import List


class Solution:
    def permuteUnique(self, nums: List[int]) -> List[List[int]]:
        ans = []
        nums.sort()
        used = [False] * len(nums)

        def dfs(depth, path):
            if depth == len(nums):
                ans.append(path)
            for i in range(len(nums)):
                if used[i] or (i > 0 and nums[i] == nums[i-1] and not used[i - 1]):
                    continue
                # path.append(nums[i])
                used[i] = True
                dfs(depth + 1, path + [nums[i]])
                # path.pop()
                used[i] = False

        dfs(0, [])
        return ans


if __name__ == "__main__":
    s = Solution()
    print(s.permuteUnique([1, 1, 2]))
